TimeSplitter methods to split events in an EventList

jmborr opened the pull request:
References [EWM 340 [Task] method(s) in DataObject::EventList to split using TimeSplitter](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=340)

**Description of work.**
- [x] `TimeSplitter` method to split an `EventList` object, and unit testing.
- [x] Auxiliary `EventList` methods for sorting and event time computation, and unit testing

**To test:**

changes not exposed to users yet. Testing restricted to unit testing.

_Release notes_ will be added in a subsequent PR when implementing this changes to algorithm `FilterEvents`


#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

jmborr committed:
fix merge conflicts

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Fix merge conflicts

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Fix merge conflicts

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
fix typo and rephrase comment

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
fix clang-format complains

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

peterfpeterson commented on Framework/DataObjects/src/TimeSplitter.cpp:
```diff
@@ -280,5 +293,171 @@ TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
+  // collect the state from events which is to be transferred to the partials
+
+  bool removeDetIDs{true};
+  auto detIDs = events.getDetectorIDs();
```
These should be `const auto &` to prevent making copies.

peterfpeterson commented on Framework/DataObjects/src/TimeSplitter.cpp:
```diff
@@ -280,5 +293,171 @@ TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
+    partial->switchTo(eventType);
+  };
+  // iterate over the partials
+  for (auto iter = cbegin(partials); iter != cend(partials); ++iter)
```
This is using constant iterators, and is exactly what [std::for_each](https://en.cppreference.com/w/cpp/algorithm/for_each) is meant to do. I think these lines should be
```c
std::for_each(partials.begin(), partials.end(), initPartial);
```

jmborr committed:
fix cppcheck whines

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
missing closing brace!

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

peterfpeterson commented on Framework/DataObjects/src/TimeSplitter.cpp:
```diff
@@ -280,5 +293,171 @@ TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
+  assert(times.size() == events.size());
+  // initialize the iterator over the splitter
+  // it assumes the splitter keys (DateAndTime objects) are sorted by increasing time.
+  auto itSpl = m_roi_map.cbegin();                // iterator over the splitter
```
The iterator variable names are a little confusing here, especially `itVec` which is an iterator for the events themselves. How about
* `itSpl` becomes `itSplitter` 
* `itTim` becomes `itTime` (one extra character helps a ton)
* `itVec` becomes `itEvent`

peterfpeterson commented on Framework/DataObjects/src/TimeSplitter.cpp:
```diff
@@ -280,5 +293,171 @@ TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
+ * @param shift : shift the TOF values after rescaling, in units of microseconds.
+ * @return : times that will be compared again the splitter boundaries, in units of nanoseconds.
+ */
+std::vector<int64_t> TimeSplitter::sortEventList(const EventList &events, bool pulseTof, bool tofCorrect, double factor,
```
I think this method could be greatly simplified by using `EventList::sort*` methods (picking the right one) and `EventList::getPulseTimes()`. I think this is correct in assuming that `EventList::sortTimeAtSample()` will give the same result as `EventList::sortPulseTime()` when `factor>=0`. The missing bit to make this "simpler" version work is an extra method `EventList::getTimesAtSample(const double &tofFactor, const double &tofOffset)` which would still return a `std::vector<DateAndTime>`.

jmborr committed:
use std::for_each and remove unused TimeROI method

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

peterfpeterson reviewed (commented):
First two questions: 
1. Why does `TimeSplitter` mess tear apart `EventList` using public `EventList`'s public info rather than `EventList` tear itself using public info from `TimeSplitter`? Since there is a lot of private state/helpers already in `EventList`, it feels like the splitting belongs over there. 
1. Why is the thing shared/compared `in64_t` rather than `DateAndTime`? You are making an extra copy of information because of this since `EventList` can create many of the vectors already.

jmborr committed:
use names for the iterators that easier to understand

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Implement EventList::getPulseTOFTimes()

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Refactor EventList::getPulseTOFTimes()

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Implement EventList::getPulseTOFTimesAtSample()

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Backend calculator EventList::eventTimesCalculator()

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Remove changes to IEventList

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Remove changes to IEventList

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
move TimeSplitter::initializePartials to EventList

This method doesn't utilize the attributes of TimeSplitter so it makes sense to move it to the EventList class

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Refactor TimeSplitter::sortEventList

This method is unnecessary

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
Use DateAndTime objects instead of int64_t numbers

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
"raises" is for Python, and "throws" is for C++

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

peterfpeterson commented on Framework/DataObjects/src/EventList.cpp:
```diff
@@ -1269,6 +1269,12 @@ MantidVec *EventList::makeDataE() const {
 
+/// Returns a copy of the Histogram associated with this spectrum.
+HistogramData::Histogram EventList::getHistogram() const {
+  HistogramData::Histogram ret(m_histogram);
```
This is return by value so you don't need to make a copy here

peterfpeterson commented on Framework/DataObjects/src/EventList.cpp:
```diff
@@ -3937,6 +3956,31 @@ void EventList::filterInPlace(Kernel::SplittingIntervalVec &splitter) {
+ * Initialize the detector ID's and event type of the destination event lists when splitting this list.
+ * @param partials : resulting partial lists of events after splitting's done
+ */
+void EventList::initializePartials(std::map<int, EventList *> partials) const {
```
`EventList` is-an `ISpectrum` so some of this can be replaced with [`ISpectrum::copyInfoFrom(other)`](https://github.com/mantidproject/mantid/blob/main/Framework/API/src/ISpectrum.cpp#L24-L29) to reduce the amount of code here.

peterfpeterson commented on Framework/DataObjects/src/TimeSplitter.cpp:
```diff
@@ -277,5 +290,122 @@ TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
+                                  bool tofCorrect, double factor, double shift) const {
+
+  if (events.getEventType() == EventType::WEIGHTED_NOTIME)
+    throw std::runtime_error("EventList::splitByTime() called on an EventList "
```
I think it would be clearer to make this [`std::invalid_argument`](https://en.cppreference.com/w/cpp/error/invalid_argument)

peterfpeterson commented on Framework/DataObjects/src/TimeSplitter.cpp:
```diff
@@ -277,5 +290,122 @@ TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
+template <typename EVENTTYPE>
+void TimeSplitter::splitEventVec(const std::vector<DateAndTime> &times, const std::vector<EVENTTYPE> &events,
+                                 std::map<int, EventList *> partials) const {
+  assert(times.size() == events.size());
```
This only gets compiled/run for debug builds. I don't know if you were relying on it for release builds.

jmborr replied:
dang it. I will change it to an `if(...){ throw(...) }`

peterfpeterson commented on Framework/DataObjects/src/TimeSplitter.cpp:
```diff
@@ -280,5 +293,171 @@ TimeROI TimeSplitter::getTimeROI(const int workspaceIndex) {
+    }
+    if (partial) {
+      const EVENTTYPE eventCopy(*itVec);
+      partial->addEventQuickly(eventCopy);
```
This still needs to be changed

peterfpeterson commented on Framework/Kernel/inc/MantidKernel/TimeROI.h:
```diff
@@ -47,6 +47,7 @@ class MANTID_KERNEL_DLL TimeROI {
   bool operator==(const TimeROI &other) const;
   /// print the ROI boundaries to a string
   std::string debugStrPrint(const std::size_t type = 0) const;
+  void debugPrint(const std::size_t type = 0) const;
```
It looks like this doesn't have an implementation since there is no change to `TimeROI.cpp`

jmborr replied:
Good catch! I thought I had gotten rid of that method.
Obliterating...

jmborr committed:
test for TimeSplitter::splitEventList

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
implement remaining review comments

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

peterfpeterson commented on Framework/DataObjects/src/EventList.cpp:
```diff
@@ -2851,45 +2854,57 @@ std::vector<double> EventList::getWeightErrors() const {
+
+/// Get the Pulse-time + TOF for each event in this EventList
+std::vector<DateAndTime> EventList::getPulseTOFTimes() const {
+  auto timeCalc = [](const auto &event) { return event.pulseTime() + static_cast<int64_t>(event.tof() * 1000.0); };
```
Please add a comment about what the 1000 is for

jmborr replied:
I introduced a `constexpr`
```cpp
constexpr double microToNano{1000.0};  // time unit conversion
```

peterfpeterson commented on Framework/DataObjects/src/EventList.cpp:
```diff
@@ -2851,45 +2854,57 @@ std::vector<double> EventList::getWeightErrors() const {
+ */
+std::vector<DateAndTime> EventList::getPulseTOFTimesAtSample(const double &factor, const double &shift) const {
+  auto timeCalc = [&](const auto &event) {
+    return event.pulseTime() + static_cast<int64_t>((factor * event.tof() + shift) * 1000.0);
```
Please add a comment about what the 1000 is for

peterfpeterson commented on Framework/DataObjects/src/EventList.cpp:
```diff
@@ -3937,6 +3953,31 @@ void EventList::filterInPlace(Kernel::SplittingIntervalVec &splitter) {
+  // collect the state from events which is to be transferred to the partials
+  bool removeDetIDs{true};
+  const auto &detIDs = this->getDetectorIDs();
+  auto histogram = this->getHistogram();
```
should `histogram` and `eventType` be declared `const` so the compiler can optimize it away?

peterfpeterson commented on Framework/DataObjects/src/EventList.cpp:
```diff
@@ -3937,6 +3953,31 @@ void EventList::filterInPlace(Kernel::SplittingIntervalVec &splitter) {
+  auto initPartial = [&](EventList *partial) {
+    partial->clear(removeDetIDs);
+    partial->copyInfoFrom(*this);
+    partial->setHistogram(histogram); // problem here!
```
Is this a spurious comment (please remove) or something that needs to be addressed/expanded on?

jmborr replied:
yes, there was an issue here some time ago but now it's fixed :smile:

jmborr committed:
implement additional review comments

Signed-off-by: Jose Borreguero <borreguero@gmail.com>

jmborr committed:
remove unused variable

Signed-off-by: Jose Borreguero <borreguero@gmail.com>
